Add indeterminate prop to the Checkbox

long-lazuli opened the pull request:
## Purpose

Use another symbol when state is indeterminate.

Compliant with design:
<img width="99" alt="image" src="https://user-images.githubusercontent.com/1178625/152643227-bd1d33c7-8c63-4225-b7dd-44505cd08d19.png">

## Approach and changes

I tried to stay as close as possible of the checked state styling implementation.
I add a new SVG in the label for the indeterminate symbol.
And I put a new props on both SVGs to be able to select them with CSS.

There is now `indeterminate` property.
It takes visually precedence visually over the checked state.
Natively, the indeterminate state of the input would be set to false, when the checked state change. 
Here, because it came from a props, it stays on.

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [x] Unit and integration tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

robinmetral reviewed (commented):
@long-lazuli thanks a lot for this, and sorry for the wait!

First of all, I like how clean this looks, after our chat on Slack I had assumed the implementation would be very JavaScript-heavy. In contrast this seems very lean 💯 

So let's get to it. First, I just wanted to make sure I understand `indeterminate` and the overall implementation correctly:

The way this would work is that since there is no `indeterminate` attribute in HTML, developers would manually set the underlying `<input type="checkbox">` element's `indeterminate` property as `true`, and this would make the `–` SVG appear via the transform.

☝️ is this right?

Overall, I really like the implementation, but I have two main concerns about it:

1. **Clarity**—I mentioned this in a comment below: since a checkbox will always have a boolean value when submitted, `indeterminate` is basically just a UI thing. Is this clear enough to users and developers? Are we sure that this couldn't lead to issues? (e.g. say an indeterminate checkbox is true under the hood, so the merchant doesn't understand that they've essentially checked it unless they explicitly untick it). Generally, wouldn't there be better UX patterns for the issue we're trying to solve here? (I'll follow-up about your use-case on Slack)
2. **Accessibility**—this is related to the above. Because this is an uncommon pattern, browser/screenreader pairs will handle the state differently. I've quickly tested with VoiceOver and on Chrome, the checkbox state is announced as "mixed", whereas it's announced on Safari as its actual state ("ticked", i.e. `true` in the story you created). I believe that this might confuse users and lead to similar issues

---

VO on Chrome:
<img width="1483" alt="Screenshot 2022-02-08 at 14 29 06" src="https://user-images.githubusercontent.com/35560568/152996924-679572eb-a73c-49d6-8002-4041dda5e663.png">

VO on Safari:
<img width="1483" alt="Screenshot 2022-02-08 at 14 28 25" src="https://user-images.githubusercontent.com/35560568/152996909-7c344822-9c74-4b35-a6fc-142666b09d69.png">

long-lazuli commented on packages/circuit-ui/components/Checkbox/Checkbox.tsx:
```diff
@@ -161,16 +163,19 @@ const inputBaseStyles = ({ theme }: StyleProps) => css`
   }
 
-  &:checked + label > svg {
+  &:checked + label > svg[data-symbol='checked'],
+  &:indeterminate + label > svg[data-symbol='indeterminate'] {
```
that's how I think about it, yes

long-lazuli commented:
You're right about webkit voiceover : https://bugs.webkit.org/show_bug.cgi?id=231760 :-/

long-lazuli commented on packages/circuit-ui/components/Checkbox/Checkbox.stories.tsx:
```diff
@@ -114,6 +119,22 @@ Disabled.args = {
+
+Indeterminate.args = {
+  name: 'indeterminate',
+  value: 'true',
```
Actually, the `value` has nothing to do with the `checked` attribute.
the `checked` attribute is just to verify if the value should be put inside `formData`.

We should just make sure that checked can be true if indeterminate is.
Or I can make the type props accept `boolean | 'mixed'` for `checked` props...

long-lazuli commented:
another exemple : https://www.w3.org/WAI/ARIA/apg/example-index/checkbox/checkbox-mixed.html
(note the use of `aria-checked="mixed"`, but no `aria-checked` so that form don't get this value in its formData on submit.)
that would be what is expected when indeterminate is true, like an unchecked one, the checkbox value should not be sent if it's indeterminate.

long-lazuli commented:
@robinmetral 
Sorry for being late to answer your concerns about Safari AT. (https://github.com/sumup-oss/circuit-ui/pull/1415#pullrequestreview-875975579)

In this PR, we don't make a new `indeterminate` props or value at all.

The checkbox state indeterminate is done by javascript, so I think it's up to the developer to use it carefully.
But I added a note on documentation about it. And added a `aria-cheked="mixed" into the concerned story.

robinmetral reviewed (commented):
Looks nice! I think I finally understand what your approach is 😅 (letting developers set `input.indeterminate` to toggle these styles). However I think I'd rather allow a new `indeterminate` prop so developers aren't forced to go get a ref and remember to also pass `aria-checked="mixed"`. What do you think? Is there any reason you didn't want to add a new prop for this?

Also we should add tests for this 🤓 (we can snapshot styles, and if we go for a prop, then we can also cover a11y by asserting that the input should have `aria-checked="mixed"` and yadda yadda)

long-lazuli commented on packages/circuit-ui/components/Checkbox/__snapshots__/Checkbox.spec.tsx.snap:
```diff
@@ -238,23 +265,34 @@ exports[`Checkbox Styles should render with default styles 1`] = `
   background-color: var(--cui-bg-accent-strong);
 }
 
+.circuit-1:checked:disabled+label::before,
+.circuit-1:checked[disabled]+label::before,
+.circuit-1:indeterminate:disabled+label::before,
+.circuit-1:indeterminate[disabled]+label::before {
+  border-color: var(--cui-border-accent-disabled);
+  background-color: var(--cui-bg-accent-strong-disabled);
+}
```
otherwise, disabled colors are applied, meaning white on white, and you can't see the checkboxes.

I added the stories to see on storybook & chromatic the 'disabled+checked', 'disabled+invalid+checked', 'invalid+indeterminate', etc...

long-lazuli commented on packages/circuit-ui/components/Checkbox/Checkbox.stories.tsx:
```diff
@@ -68,9 +66,26 @@ export const Disabled = (args: CheckboxProps) => (
+        // eslint-disable-next-line no-param-reassign
+        el.indeterminate = true;
+      }
+    }}
```
This PR was just about making the style of the Checkbox as useful as the browser checkbox,

I don't know how about the way we would like to use it.
So I just used the native behavior, allowing the `:indeterminate` state to show up.

robinmetral replied:
Got it 👍 

This is how they do it over at Spectrum:

https://github.com/adobe/react-spectrum/blob/b64e64f28b3d06a69ad2c5ae0dd994c3a07d4099/packages/%40react-aria/checkbox/src/useCheckbox.ts#L42-L49

I kind of like this because it abstracts the complexity of "yeah a checkbox can be indeterminate but you _have to_ use JS for it, and in React that means you need to use the ref to the underlying HTML checkbox node" into the design system, and then developers could just set a boolean prop `isIndeterminate` or `indeterminate` to get the styles.

Happy to merge this though (since as you said, this was always meant for styles) and then make a follow-up PR to add the convenience prop. Let me know what you prefer (in this PR or as follow-up) 🙂

long-lazuli replied:
It is, here too, stated that indeterminism is presentational only.
https://github.com/adobe/react-spectrum/blob/b64e64f28b3d06a69ad2c5ae0dd994c3a07d4099/packages/%40react-types/checkbox/src/index.d.ts#L67
As we thought, it is the responsability of the developper to handle the checked state according to its business logic.

Therefore, I'm not sure the property should be available like the checked one...

long-lazuli commented:
I do think this is sufficient for most of developpers, I'm not feeling good making a props for making the indeterminate state as, like you said, it's something that would not lead to clarity.

But one good point about this, would be to put the `aria-checked="mixed"` automatically.

What would be the property ?
`showIndeterminate: boolean` ?

long-lazuli commented:
( sorry, I think I miss-clicked, didn't want to close this PR. )

long-lazuli commented:
We now have a `indeterminate` props.

long-lazuli commented:
[update]
I made the changes, but for the moment, I see a bug (in storybook) where the indeterminate state is not working.
![image](https://user-images.githubusercontent.com/1178625/226500640-3b3192cd-3cc0-466f-b485-cc85b24149ab.png)

Still investigating this.

connor-baer commented:
@long-lazuli I've refactored large parts of the Checkbox component. I hope you don't mind if I rebase your branch to resolve the conflicts. I'd love to push this across the finish line.
